### PR TITLE
dom: Add types progressively to dom modules

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds more types to `dom` progressively, starting with just three modules.

Also adds helper that @sirreal suggested, `assertIsDefined` which covers the various null cases we were seeing without adding null checks everywhere. These null checks _are necessary_ and the original code did not handle them, which is dangerous.

## How has this been tested?
Unit tests pass and type checks pass.

## Types of changes
Adds types and fixes unhandle null checks.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/dom/src/utils/assert-is-defined.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export function assertIsDefined< T >(
```
Comment (tyxla) [inline packages/dom/src/utils/assert-is-defined.ts:1]: Do we need to add any tests for this one? Could be testing `assertIsDefined` specifically or the code that uses it.

Comment (sarayourfriend) [reply packages/dom/src/utils/assert-is-defined.ts:1]: Sure!

**Review by tyxla — COMMENTED** [review]
This looks good to me 👍 

Any idea if the test failures can be related for some reason?

**Comment by sarayourfriend** [comment]
I'm not sure. I've rebased to see if it fixes anything.

**Comment by sarayourfriend** [comment]
Seems like they might be related. I'll investigate further.

**Comment by sarayourfriend** [comment]
@tyxla this is ready for another review.

**Review by tyxla — APPROVED** [review]
This looks good to me 👍

**Pull request merged**